src/app/script/tabs_widget.cpp

```diff
@@ -0,0 +1,166 @@
+
+#ifdef ENABLE_UI
+
+using namespace ui;
+
+namespace app {
+namespace script {
+
+// static
+ui::WidgetType Tabs::Type()
+{
+  static ui::WidgetType type = ui::kGenericWidget;
+  if (type == ui::kGenericWidget)
+    type = ui::register_widget_type();
+  return type;
+}
+
+Tabs::Tabs(int selectorFlags) : m_selector(new HBox()),
+                                m_sepL(separator()),
+                                m_sepR(separator()),
+                                m_sepTB(separator()),
+                                m_buttonsBox(new HBox()),
+                                m_pages(new VBox()),
+                                m_selectorFlags(selectorFlags)
+{
+  setType(Type());
+  layoutChilds();
+}
+
+app::ButtonSet::Item* Tabs::addTab(Grid* content)
+{
+  m_pages->addChild(content);
+
+  if (!m_buttons) {
+    m_buttons = new ButtonSet(1);
+    m_buttonsBox->addChild(m_buttons);
+    m_buttons->ItemChange.connect([this](ButtonSet::Item* selItem) {
+      int oldSelectedTabIndex = m_selectedTab;
+      for (int i=0; i<m_pages->children().size(); ++i) {
+        auto tab = m_pages->children()[i];
+        bool isSelectedTab = selItem->text() ==  tab->text();
+        tab->setVisible(isSelectedTab);
+        if (isSelectedTab)
+          m_selectedTab = i;
+      }
+      layout();
+      if (oldSelectedTabIndex != m_selectedTab) {
+        std::string deselTabId = tabId(oldSelectedTabIndex);
+        std::string selTabId = tabId(m_selectedTab);
+        TabChanged(deselTabId, selTabId);
+      }
+    });
+  }
+  else {
+    m_buttons->setColumns(m_pages->children().size());
+  }
+  auto item = m_buttons->addItem(content->text());
+  // Select the first tab by default.
+  if (m_buttons->children().size() == 1) {
+    selectTab(0);
+  }
+
+  m_buttons->initTheme();
+
+  return item;
+}
+
+void Tabs::selectTab(int index)
+{
+  if (index >= 0 && index < m_pages->children().size()) {
+    for (int i=0; i<m_pages->children().size(); ++i) {
+      m_buttons->getItem(i)->setSelected(i == index);
+      m_pages->children()[i]->setVisible(i == index);
+    }
+    m_selectedTab = index;
+  }
+}
+
+void Tabs::setSelectorFlags(int selectorFlags)
+{
+  if (selectorFlags != m_selectorFlags) {
+    m_selectorFlags = selectorFlags;
+    layoutChilds();
+  }
+}
+
+int Tabs::tabIndexById(std::string id) const
+{
+  for (int i=0; i<m_pages->children().size(); ++i) {
+    if (m_pages->children()[i]->id() == id)
+      return i;
+  }
+  return -1;
+}
+
+int Tabs::tabIndexByText(std::string text) const
+{
+  for (int i=0; i<m_pages->children().size(); ++i) {
+    if (m_pages->children()[i]->text() == text)
+      return i;
+  }
+  return -1;
+}
+
+std::string Tabs::tabId(int index) const
+{
+  return index >= 0 && index < m_pages->children().size() ? m_pages->children()[index]->id() : "";
+}
+
+std::string Tabs::tabText(int index) const
+{
+  return index >= 0 && index < m_pages->children().size() ? m_pages->children()[index]->text() : "";
+}
```

Comment:
gcc and clang optimize this correctly, but msvc [prefers std::string()](https://godbolt.org/z/WvEToo5f7) instead [of ""](https://godbolt.org/z/eqovn8xGT) for an empty string. And break the length of these lines:
```suggestion
std::string Tabs::tabId(int index) const
{
  return index >= 0 && index < m_pages->children().size() ? m_pages->children()[index]->id() : std::string();
}

std::string Tabs::tabText(int index) const
{
  return index >= 0 && index < m_pages->children().size() ? m_pages->children()[index]->text() : std::string();
}
```